File: src/FDtomo.c
#include "common/read_spec.h"
#include "common/velocity_model.h"
#include "common/station.h"
#include "FDtomo/make1d.h"
#include "FDtomo/c2f.h"
#include "FDtomo/sphfd.h"
#include "FDtomo/sphfdloc.h"
#include "FDtomo/sphrayderv.h"
#include "FDtomo/runlsqr.h"
#include "FDtomo/makenewmod.h"
#include "common/grid.h"
#include "common/shared_variables.h"
#include "common/vec/vec.h"
#include <stdio.h>
int main(int argc, char *argv[]){
	
	char spec_path[MAXSTRLEN] = "../data/small/FDtomo01.spec";
	char *model1D_path = "../data/small/TW_m30_mdl";
	char *stafile = "../data/small/runs_files/stationloc_out.txt";
	char *leqsfil = "../data/small/runs_files/arrivals/All.txt";
	
	/*
	char spec_path[MAXSTRLEN] = "../data/HL_2km/run_spec_file01";
	char *model1D_path = "../data/HL_2km/CWB_mdl";
	char *stafile = "../data/HL_2km/run_files/HL171819.coords";
	char *leqsfil = "../data/HL_2km/run_files/arrivals/hyp.Fdtomo171819";
	*/

	CommonEnv common_env = setCommonEnv(spec_path);
	LocEnv loc_env = setLocEnv(spec_path);
	SphraydervEnv sphrayderv_env = setSphraydervEnv(spec_path);
	RunlsqrEnv runlsqr_env = setRunlsqrEnv(spec_path);
	MakenewmodEnv makenewmod_env = setMakeNewmodEnv(spec_path);
	Mesh3D mesh = setMesh3D(spec_path);
	Coordinate3D coarseCoordinate =  setCoordinate(spec_path);
	Coordinate3D fineCoordinate = setCoordinate(spec_path);

	Mesh3D fineMesh = generateFineMesh(mesh);
	copyMesh3D(&fineCoordinate.mesh, &fineMesh);
	
	velocityModel1D vpModel; 
	velocityModel1D vsModel; 
	char interp[MAX1D];
	readVelocityModel1D(model1D_path, &vpModel, &vsModel, interp);
	
	Mesh1D coarseMesh = createMesh1D(coarseCoordinate.mesh.numberOfNode.z, coarseCoordinate.mesh.gridz);
	Coordinate1D coarseCoordinate1D =  createCoordinate(coarseMesh, coarseCoordinate.space.z, coarseCoordinate.origin.z);
	
	velocityModel1D newVpModel = transform1D(coarseCoordinate1D, vpModel, interp);
	velocityModel3D coarseVpModel = create3DModel(coarseCoordinate, newVpModel);

	EventNode *event_list = createEventList(leqsfil);
	StationNode *station_list = createStationList(stafile, 1);
	int table_size = getStationCount(station_list);
	Station *station_array = StationList2Arr(station_list);
	Event *event_array = EventList2Arr(event_list);

	for(int i = 0; i < 1; i++){
		velocityModel3D fineVpModel = transform3D(fineCoordinate, coarseVpModel);
		
		velocityModel1D newVsModel = transform1D(coarseCoordinate1D, vsModel, interp);
		velocityModel3D coarseVsModel = create3DModel(coarseCoordinate, newVsModel);
		velocityModel3D fineVsModel = transform3D(fineCoordinate, coarseVsModel);

		travelTimeTable *vpTable_array = sphfdAll(fineVpModel, fineVsModel, station_array, table_size);
		
		int event_size = getEventCount(event_list);
		EventNode *new_event_list = sphfdloc(fineCoordinate, vpTable_array, table_size, event_array, event_size, loc_env);

		int new_event_size = getEventCount(new_event_list);
		Event *new_event_array = EventList2Arr(new_event_list);
		SPHRAYDERV_DATA *SPHRAYDERV = sphrayderv(coarseVpModel, vpTable_array, new_event_array, new_event_size, station_array, table_size, sphrayderv_env, common_env);

		RUNLSQR_DATA *RUNLSQR = runlsqr(SPHRAYDERV, runlsqr_env, common_env);
		makenewmod(coarseCoordinate, &coarseVpModel, &coarseVsModel, RUNLSQR, table_size, makenewmod_env, common_env);
		free(SPHRAYDERV);
		free(RUNLSQR);
		free(new_event_array);
		free(vpTable_array);
	}
	return 0;

}

File: src/pythonModules/FDtomo.py
import coordinate, velocity_model, environment, station, travel_time, event
from velocity_model import VelocityModel1D, VelocityModel3D
from mesh import Mesh1D, Mesh3D
from station import Station
from event import Event
import _FDtomoC
import abc
from tomography import *
file_path = "../../data/small/FDtomo01.spec"
model1D_path = "../../data/small/TW_m30_mdl"
stafile = "../../data/small/runs_files/stationloc_out.txt"
leqsfil = "../../data/small/runs_files/arrivals/All.txt"

loc_env = environment.LocEnv().create(file = file_path)
sphrayderv_env = environment.SphraydervEnv().create(file = file_path)
runlsqr_env = environment.RunlsqrEnv().create(file = file_path)
makenewmod_env = environment.MakenewmodEnv().create(file = file_path)

environment = {'loc_env' : loc_env, 'sphrayderv_env' : sphrayderv_env, 'runlsqr_env' : runlsqr_env, 'makenewmod_env' : makenewmod_env}

coarseMesh3D = Mesh3D().create(file = file_path)
fineMesh3D = coarseMesh3D.generateFineMesh()

event = Event().createArray(leqsfil)
station = Station().createArray(file = stafile)

vpModel = VelocityModel1D()
vsModel = VelocityModel1D()
VelocityModel1D().setVelocityModel(model1D_path, vpModel, vsModel)

TomographyBuilder() \
    .Environment(environment) \
    .Event(event) \
    .Station(station) \
    .VelocityModel() \
        .Coordinate() \
            .Mesh(coarseMesh3D, fineMesh3D) \
            .Origin([120.9, 23.8, -4.0]) \
            .Space([2,2,2]) \
        .ReferenceModel(vpModel, vsModel) \
    .execute() 

'''
loc_env = environment.LocEnv().create(file = file_path)
sphrayderv_env = environment.SphraydervEnv().create(file = file_path)
runlsqr_env = environment.RunlsqrEnv().create(file = file_path)
makenewmod_env = environment.MakenewmodEnv().create(file = file_path)


mesh3D = Mesh3D().create(file = file_path)
coarseCoordinate3D = coordinate.Coordinate3D().create(file = file_path)
fineCoordinate3D = coordinate.Coordinate3D().create(file = file_path)
fineMesh3D = mesh3D.generateFineMesh()
fineCoordinate3D.setMesh(fineMesh3D)

vpModel = velocity_model.VelocityModel1D()
vsModel = velocity_model.VelocityModel1D()
velocity_model.VelocityModel1D().setVelocityModel(model1D_path, vpModel, vsModel)

#tmp = [2 for i in range(21)]
#tmp[1:12] = [1 for i in range(11)]
tmp = [2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2]
fineMesh1D = mesh.Mesh1D().create(numberOfNode = 22, igrid = tmp)
fineCoordinate1D = coordinate.Coordinate1D().create(fineMesh1D, 2, -4)


fineVpModel = vpModel.transform(fineCoordinate1D)
fineVsModel = vsModel.transform(fineCoordinate1D)

CoarseVpModel3D = velocity_model.VelocityModel3D().create(coarseCoordinate3D, fineVpModel)
fineVpModel3D = CoarseVpModel3D.transform(fineCoordinate3D)

CoarseVsModel3D = velocity_model.VelocityModel3D().create(coarseCoordinate3D, fineVsModel)


station_array = station.Station().createArray(file = stafile)
table_array = []
for i in range(len(station_array)):
   table = travel_time.TravelTimeTable().create(fineVpModel3D, station_array[i])
   table_array.append(table)

event_array = event.Event().createArray(leqsfil)
#for i in range(len(event_array)):
 #   new_event = event.Event().singleLoc(fineCoordinate3D, table_array, event_array[i], loc_env)
  #  new_event_array.append(new_event)

new_event_array = event.Event().sphlocating(fineCoordinate3D, table_array, loc_env, leqsfil)

event_size = len(new_event_array)
table_size = len(table_array)
derv, residual_vector = event.Event().sphRaytracing(CoarseVpModel3D, table_array, new_event_array, event_size, station_array, table_size, sphrayderv_env)
perturbation = event.Event().runlsqr(derv, residual_vector, runlsqr_env)
velocity_model.VelocityModel3D().makeNewModel(coarseCoordinate3D, CoarseVpModel3D, CoarseVsModel3D, perturbation, table_size, makenewmod_env)
'''